### PR TITLE
Demonstrate static checking with curl_handles_lock

### PR DESCRIPTION
**Pull request by gaul**
This can find missing locks at compile-time.

### CODE REVIEW

**Comment by gaul** [comment]
@ggtakec I'd like to move towards static lock checking instead of relying on only dynamic ThreadSanitizer.  I think this can find bugs more reliably.  What do you think?  `FdEntity` will be a more complicated case but I can demonstrate this before moving any call sites.

**Comment by gaul** [comment]
Also the `Mutex` wrappers should not be required but we would need to opt into clang libc++ instead of GNU libstdc++.

File: .github/workflows/linux-ci-helper.sh [hunk 1]
```diff
@@ -290,8 +293,11 @@ fi
 #-----------------------------------------------------------
 echo "${PRGNAME} [INFO] Set environment for configure options"
 
-echo "CXXFLAGS=${CXXFLAGS}"                   >> "${GITHUB_ENV}"
-echo "CONFIGURE_OPTIONS=${CONFIGURE_OPTIONS}" >> "${GITHUB_ENV}"
+cat << EOF > "${GITHUB_ENV}"
```
Comment (ggtakec) [inline .github/workflows/linux-ci-helper.sh:296]: The ${GITHUB_ENV} file has been overwritten.
Please change to addendum.
Is there a reason why you changed to Here Document?

File: src/mutex.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
+/*
```
Comment (ggtakec) [inline src/mutex.h:1]: Please make sure to indent 4 space characters.

**Review by ggtakec — COMMENTED** [review]
I think I can promote this amendment.

I have a few questions, so I will summarize them here.
##### About license
Is there any external license involved in mutex.h?
I was curious about this because when I searched for `THREAD_ANNOTATION_ATTRIBUTE__`, I saw some examples of like code.

##### About AutoLock::Type
This `Mutex` class will replace all `AutoLock` classes, but can't it implement `AutoLock::Type`?
(Is it difficult to add attributes for `ThreadSanitizer` ?)
However, there is no particular problem even if it cannot be implemented.

**Comment by gaul** [comment]
Sorry this should be broken into separate PRs:

* Remove `ALREADY_LOCKED` state
* Move to `std::mutex`
* Enable static lock checking

**Comment by gaul** [comment]
Pursing this in a variety of separate PRs.

**Pull request closed without merging**